### ISSUE TITLE
图片背景预测结果是0.5的原因是什么

### ISSUE THREAD

**Issue by liang23333** [fengjian0106/hed-tutorial-for-document-scanning#12]
作者您好，我在实现边缘检测网络时遇到了背景色全是灰色的情况，我查了一下，背景像素的预测结果全是0.5，请问造成这种情况的原因可能是什么呢？

**Comment by liang23333** [comment]
还有就是在训练的过程中，loss几乎不收敛，一直在某个范围内震荡，这是什么原因造成的呢？

**Comment by fengjian0106** [comment]
你用的是自己的训练样本图片吗？
可以先使用少量的图片，比如1张或10张做训练，如果 loss 不收敛，检查一下你的图片预处理流程和我提供的是否是一致的，训练的每一个 step 里面，也可以把 label 数据、output 数据都绘制出来看效果

**Comment by liang23333** [comment]
作者您好，感谢您的回复。我先前没有说清楚，我训练的是自己的网络，改动了HED的网络架构，但是基本逻辑不变。  图片预处理的话我是X-mean，Y把大于某个阈值的数据设置为1，小于阈值的数据设置为0。 我想请问一下您训练的时候loss的收敛是如何收敛的呢，从哪个值到哪个值呢，我的loss一直在1左右震荡。

**Comment by fengjian0106** [comment]
我的loss就是比较正常的那种，从大于1的数值，降低到0.00XX这种形式的，小数点后有几个0我记不清楚了。你应该检查一下你的网络结构，用少量的图片先做尝试，看看 loss 是否可以收敛到一个比较合理的取值范围内。网络结构如果本身就有问题，那 loss 是有可能不正常的，甚至会出现发散的情况

**Comment by liang23333** [comment]
作者您好，我是复现的RCF edge detection的网络，github上有他的caffe版本的源码，所以网络架构上是有保证的。所以我猜测是参数或者其他的一些原因导致网络无法收敛，背景点的预测一直是0.5，但是边缘是可以检测出来的。

**Comment by fengjian0106** [comment]
导致不收敛的原因是有很多种可能的，我不清楚你的网络结构具体是怎样的，任何一个微小的差异都可能会引起一些意料之外的情况，在我的个人经验里，遇到的大多数的不收敛，都是网络结构引起的。如果是超参数的话，learning rate 太大，可能会引起 loss 的震荡，但是在起始阶段通常也会看到loss的值有一个很明显的下降过程。

**Comment by liang23333** [comment]
请问正确的输入数据的预处理方法应该是怎么样呢？对于原图我是cv2.imread然后直接减去(104.00699, 116.66877, 122.67892)   然后GT我是imread(path,0)然后再对大于某个阈值的设置为1，其他的设置为0。请问这个过程是正确的吗

**Comment by liang23333** [comment]
我今天尝试了HED的网络架构，完整复现的话，如果不考虑中间每层的1e-6或者1e-7，收敛的过程也非常缓慢，，而且最终也不清楚会收敛到哪里去？不过比较奇怪的是论文中为了保持图片大小，好像是deconv之后然后crop成输入图片的大小，所以batch得大小只能是1。我复现的时候也是类似的，VGG架构采用的是https://github.com/machrisaa/tensorflow-vgg, deconv和loss的计算方式我用的是您的方式，不同的是为了保持和HED的caffe代码的一致性，我是单张图片进行运算的，但还是无法收敛，https://github.com/liang23333/edge_detection，请问您能给我一些建议吗？

**Comment by liang23333** [comment]
`import tensorflow as tf

import numpy as np
from functools import reduce

VGG_MEAN = [103.939, 116.779, 123.68]


class HED:
    """
    A trainable version VGG19.
    """

    def __init__(self, vgg19_npy_path='vgg16.npy', trainable=True, dropout=0.5):
        if vgg19_npy_path is not None:
            self.data_dict = np.load(vgg19_npy_path, encoding='latin1').item()
        else:
            self.data_dict = None

        self.var_dict = {}
        self.trainable = trainable
        self.dropout = dropout
        self.image = tf.placeholder(tf.float32, [1, None, None, 3], name='input_image')
        self.label = tf.placeholder(tf.float32, [1, None, None, 1], name='input_label')
        """
        load variable from npy to build the VGG
        :param rgb: rgb image [batch, height, width, 3] values scaled [0, 1]
        :param train_mode: a bool tensor, usually a placeholder: if True, dropout will be turned on
        """

        # rgb_scaled = rgb * 255.0
        #
        # # Convert RGB to BGR
        # red, green, blue = tf.split(axis=3, num_or_size_splits=3, value=rgb_scaled)
        # assert red.get_shape().as_list()[1:] == [224, 224, 1]
        # assert green.get_shape().as_list()[1:] == [224, 224, 1]
        # assert blue.get_shape().as_list()[1:] == [224, 224, 1]
        # bgr = tf.concat(axis=3, values=[
        #     blue - VGG_MEAN[0],
        #     green - VGG_MEAN[1],
        #     red - VGG_MEAN[2],
        # ])
        # assert bgr.get_shape().as_list()[1:] == [224, 224, 3]

        self.conv1_1 = self.conv_layer(self.image, 3, 64, "conv1_1")
        self.conv1_2 = self.conv_layer(self.conv1_1, 64, 64, "conv1_2")
        self.pool1 = self.max_pool(self.conv1_2, 'pool1')

        self.conv2_1 = self.conv_layer(self.pool1, 64, 128, "conv2_1")
        self.conv2_2 = self.conv_layer(self.conv2_1, 128, 128, "conv2_2")
        self.pool2 = self.max_pool(self.conv2_2, 'pool2')

        self.conv3_1 = self.conv_layer(self.pool2, 128, 256, "conv3_1")
        self.conv3_2 = self.conv_layer(self.conv3_1, 256, 256, "conv3_2")
        self.conv3_3 = self.conv_layer(self.conv3_2, 256, 256, "conv3_3")
        self.pool3 = self.max_pool(self.conv3_3, 'pool3')

        self.conv4_1 = self.conv_layer(self.pool3, 256, 512, "conv4_1")
        self.conv4_2 = self.conv_layer(self.conv4_1, 512, 512, "conv4_2")
        self.conv4_3 = self.conv_layer(self.conv4_2, 512, 512, "conv4_3")
        self.pool4 = self.max_pool(self.conv4_3, 'pool4')

        self.conv5_1 = self.conv_layer(self.pool4, 512, 512, "conv5_1")
        self.conv5_2 = self.conv_layer(self.conv5_1, 512, 512, "conv5_2")
        self.conv5_3 = self.conv_layer(self.conv5_2, 512, 512, "conv5_3")

        #first

        self.conv1_end=self.conv_layer(self.conv1_2,64,1,'conv1_end',kernel_size=1)
        self.dsn1=self.conv1_end
        self.out1=tf.sigmoid(self.dsn1)
        self.loss1=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn1,label=self.label)

        #second

        self.conv2_end=self.conv_layer(self.conv2_2,128,1,'conv2_end',kernel_size=1)
        self.dsn2=self.deconv(self.conv2_end,2)
        self.out2=tf.sigmoid(self.dsn2)
        self.loss2=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn2,label=self.label)

        #third
        self.conv3_end=self.conv_layer(self.conv3_3,256,1,'conv3_end',kernel_size=1)
        self.dsn3=self.deconv(self.conv3_end,4)
        self.out3=tf.sigmoid(self.dsn3)
        self.loss3=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn3,label=self.label)

        #fourth
        self.conv4_end=self.conv_layer(self.conv4_3,512,1,'conv4_end',kernel_size=1)
        self.dsn4=self.deconv(self.conv4_end,8)
        self.out4=tf.sigmoid(self.dsn4)
        self.loss4=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn4,label=self.label)


        #fifth
        self.conv5_end=self.conv_layer(self.conv5_3,512,1,'conv5_end',kernel_size=1)
        self.dsn5=self.deconv(self.conv5_end,16)
        self.out5=tf.sigmoid(self.dsn5)
        self.loss5=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn5,label=self.label)


        #concat

        self.Concat=tf.concat([self.dsn1,self.dsn2,self.dsn3,self.dsn4,self.dsn5],axis=3,name='Concat')
        self.dsn=self.conv_layer(self.Concat,5,1,'dsn',kernel_size=1)
        self.out=tf.sigmoid(self.dsn)
        self.loss=self.class_balanced_sigmoid_cross_entropy(logits=self.dsn,label=self.label)


        self.loss_sum=self.loss1+self.loss2+self.loss3+self.loss4+self.loss5+self.loss
        tf.summary.scalar('loss_sum',self.loss_sum)
    def class_balanced_sigmoid_cross_entropy(self,logits, label):
        ## ref - https://github.com/ppwwyyxx/tensorpack/blob/master/examples/HED/hed.py
        """
        The class-balanced cross entropy loss,
        as in `Holistically-Nested Edge Detection
        <http://arxiv.org/abs/1504.06375>`_.
        Args:
            logits: of shape (b, ...).
            label: of the same shape. the ground truth in {0,1}.
        Returns:
            class-balanced cross entropy loss.
        """

        with tf.name_scope('class_balanced_sigmoid_cross_entropy'):
            count_neg = tf.reduce_sum(1.0 - label)  # 样本中0的数量
            count_pos = tf.reduce_sum(label)  # 样本中1的数量(远小于count_neg)
            # print('debug, ==========================, count_pos is: {}'.format(count_pos))
            beta = count_neg / (count_neg + count_pos)  ## e.g.  60000 / (60000 + 800) = 0.9868

            pos_weight = beta / ((1.0 - beta)*2)  ## 0.9868 / (1.0 - 0.9868) = 0.9868 / 0.0132 = 74.75
            cost = tf.nn.weighted_cross_entropy_with_logits(logits=logits, targets=label, pos_weight=pos_weight)
            print(cost.shape)
            cost = tf.reduce_mean(cost * ((1 - beta)*2))

            # 如果样本中1的数量等于0，那就直接让 cost 为 0，因为 beta == 1 时， 除法 pos_weight = beta / (1.0 - beta) 的结果是无穷大
            zero = tf.equal(count_pos, 0.0)
            final_cost = tf.where(zero, 0.0, cost)
        return final_cost

    def get_kernel_size(self,factor):
        """
        Find the kernel size given the desired factor of upsampling.
        """
        return 2 * factor - factor % 2

    def upsample_filt(self,size):
        """
        Make a 2D bilinear kernel suitable for upsampling of the given (h, w) size.
        """
        factor = (size + 1) // 2
        if size % 2 == 1:
            center = factor - 1
        else:
            center = factor - 0.5
        og = np.ogrid[:size, :size]
        return (1 - abs(og[0] - center) / factor) * (1 - abs(og[1] - center) / factor)

    def bilinear_upsample_weights(self,factor, number_of_classes):
        """
        Create weights matrix for transposed convolution with bilinear filter
        initialization.
        """
        filter_size = self.get_kernel_size(factor)

        weights = np.zeros((filter_size,
                            filter_size,
                            number_of_classes,
                            number_of_classes), dtype=np.float32)

        upsample_kernel = self.upsample_filt(filter_size)

        for i in range(number_of_classes):
            weights[:, :, i, i] = upsample_kernel

        return weights


    def deconv(self,inputs, upsample_factor):

        # Calculate the ouput size of the upsampled tensor

        upsample_filter_np = self.bilinear_upsample_weights(upsample_factor, 1)
        upsample_filter_tensor = tf.Variable(upsample_filter_np)

        # Perform the upsampling
        upsampled_inputs = tf.nn.conv2d_transpose(inputs, upsample_filter_tensor,
                                                  output_shape=tf.shape(self.label),
                                                  strides=[1, upsample_factor, upsample_factor, 1])

        return upsampled_inputs




    def avg_pool(self, bottom, name):
        return tf.nn.avg_pool(bottom, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME', name=name)

    def max_pool(self, bottom, name):
        return tf.nn.max_pool(bottom, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME', name=name)

    def conv_layer(self, bottom, in_channels, out_channels, name,kernel_size=3):
        with tf.variable_scope(name):
            filt, conv_biases = self.get_conv_var(kernel_size, in_channels, out_channels, name)

            conv = tf.nn.conv2d(bottom, filt, [1, 1, 1, 1], padding='SAME')
            bias = tf.nn.bias_add(conv, conv_biases)
            relu = tf.nn.relu(bias)

            return relu

    def fc_layer(self, bottom, in_size, out_size, name):
        with tf.variable_scope(name):
            weights, biases = self.get_fc_var(in_size, out_size, name)

            x = tf.reshape(bottom, [-1, in_size])
            fc = tf.nn.bias_add(tf.matmul(x, weights), biases)

            return fc

    def get_conv_var(self, filter_size, in_channels, out_channels, name):
        initial_value = tf.truncated_normal([filter_size, filter_size, in_channels, out_channels], 0.0, 0.001)
        filters = self.get_var(initial_value, name, 0, name + "_filters")

        initial_value = tf.truncated_normal([out_channels], .0, .001)
        biases = self.get_var(initial_value, name, 1, name + "_biases")

        return filters, biases

    def get_fc_var(self, in_size, out_size, name):
        initial_value = tf.truncated_normal([in_size, out_size], 0.0, 0.001)
        weights = self.get_var(initial_value, name, 0, name + "_weights")

        initial_value = tf.truncated_normal([out_size], .0, .001)
        biases = self.get_var(initial_value, name, 1, name + "_biases")

        return weights, biases

    def get_var(self, initial_value, name, idx, var_name):
        if self.data_dict is not None and name in self.data_dict:
            value = self.data_dict[name][idx]
        else:
            value = initial_value

        if self.trainable:
            var = tf.Variable(value, name=var_name)
        else:
            var = tf.constant(value, dtype=tf.float32, name=var_name)

        self.var_dict[(name, idx)] = var

        # print var_name, var.get_shape().as_list()
        assert var.get_shape() == initial_value.get_shape()

        return var

    def save_npy(self, sess, npy_path="./vgg16-save.npy"):
        assert isinstance(sess, tf.Session)

        data_dict = {}

        for (name, idx), var in list(self.var_dict.items()):
            var_out = sess.run(var)
            if name not in data_dict:
                data_dict[name] = {}
            data_dict[name][idx] = var_out

        np.save(npy_path, data_dict)
        print(("file saved", npy_path))
        return npy_path

    def get_var_count(self):
        count = 0
        for v in list(self.var_dict.values()):
            count += reduce(lambda x, y: x * y, v.get_shape().as_list())
        return count`

**Comment by liang23333** [comment]
连接出问题了，网络架构代码如上所示。

**Comment by fengjian0106** [comment]
你把 out1 ~ out5，out 都画出来(可以参考 plot_and_save_image 函数的做法)，看看是什么样式的内容。
另外，可以看一下我提供的 def vgg_style_hed(inputs, batch_size, is_training) 函数，再对照一下我的博客里面的『矩阵初始化』和『转置卷积/反卷积的初始化』这两个章节。

**Comment by fangxu622** [comment]
@fengjian0106 方便分享以下第二步制作的数据集合吗 ？不是mac ,想用你的数据做一下参考···再制作自己的数据集
